Add missing documentation to crypto functions

markopiers opened the pull request:
## Description

This PR adds documentation to multiple functions in the `openzeppelin-crypto` crate that were missing proper documentation as described in issue #519. The additions follow the existing documentation style throughout the codebase.

### Changes

Added documentation to:

#### In `poseidon2/mod.rs`:
- Added detailed explanations for all round-related functions 
- Documented internal/external round operations (`partial_round_start`, `partial_round_end`, etc.)
- Added documentation to matrix multiplication functions
- Documented S-box application and round constant addition

#### In `curve/mod.rs`:
- Added documentation to batch inversion utilities
- Explained the Montgomery trick implementation for efficient batch inversion

#### In `hash.rs`:
- Improved documentation for `hash_pair` function with clear parameter descriptions

#### Added missing documentation for `pub(crate)` functions:
- `non_empty_u8_vec_strategy()` in test_helpers.rs
- `parse_utf8_byte()` in arithmetic/uint.rs

All documentation follows the existing style with clear descriptions, explanations of algorithms where relevant, parameter documentation, and return value descriptions.

<!--
Thank you for your interest in contributing to OpenZeppelin!

Consider opening an issue for discussion prior to submitting a PR. New features will be merged faster if they were first discussed and designed with the team.

Describe the changes introduced in this pull request. Include any context necessary for understanding the PR's purpose.
-->

<!-- Fill in with issue number -->
Resolves #519

#### PR Checklist

<!--
Before merging the pull request all of the following must be completed.
Feel free to submit a PR or Draft PR even if some items are pending.
Some of the items may not apply.
-->

- [ ] Tests
- [x] Documentation
- [ ] Changelog

markopiers commented:
@0xNeshi @bidzyyys @qalisander can someone please approve the ci tests?

0xNeshi commented on lib/crypto/src/arithmetic/uint.rs:
```diff
@@ -789,7 +789,18 @@ const fn parse_digit(utf8_digit: u8, digit_radix: u32) -> u32 {
 }
 
-/// Parse a single UTF-8 byte.
+/// Parse a single UTF-8 byte into a char.
+///
+/// Converts bytes to characters during compile-time string evaluation.
+/// Only handles ASCII bytes (0x00-0x7F) and panics on non-ASCII bytes.
+///
+/// # Arguments
+///
+/// * `byte` - Byte to convert
+///
+/// # Panics
+///
+/// On non-ASCII bytes (>= 0x80)
```
```suggestion
/// Parse a single UTF-8 byte into a char.
///
/// Converts bytes to characters during compile-time string evaluation.
/// Only handles ASCII bytes (0x00-0x7F).
///
/// # Arguments
///
/// * `byte` - Byte to convert.
///
/// # Panics
///
/// * If the byte is non-ASCII (>= 0x80).
```
nit: existing docs convention - use lists where possible, all list items end with dots.

You don't have to spend too much time improving existing documentation for `pub(crate)` and private items, for those it should be enough to have a clear sentence explaining the purpose of the item. 
On the other hand, `pub` items should be properly documented, as you've done here :+1:

0xNeshi commented on lib/crypto/src/curve/mod.rs:
```diff
@@ -236,15 +236,26 @@ pub trait AffineRepr:
+///
+/// # Arguments
+///
+/// * `v` - Mutable slice of field elements for in-place inversion
```
```suggestion
/// * `v` - Mutable slice of field elements for in-place inversion.
```
nit: add dots to ends of list items for consistency with other existing docs

0xNeshi commented on lib/crypto/src/hash.rs:
```diff
@@ -127,6 +127,16 @@ pub trait BuildHasher {
 }
 
 /// Hash the pair `(a, b)` with `state`.
+///
+/// # Arguments
+///
+/// * `a` - The first value to hash
+/// * `b` - The second value to hash 
+/// * `state` - The hasher state to use
+///
+/// # Returns
+///
+/// The finalized hash output from the hasher
```
```suggestion
/// Hash the pair `(a, b)` with `state`.
///
/// Returns the finalized hash output from the hasher.
///
/// # Arguments
///
/// * `a` - The first value to hash
/// * `b` - The second value to hash 
/// * `state` - The hasher state to use
```
nit: current convention is we omit the `# Returns` section and include it in the general description.

0xNeshi reviewed (commented):
Loving the docs :+1: Once you address these minor comments, we can merge

markopiers commented:
hey @0xNeshi thanks for the feedback, I commited the changes you suggested. I think this could be merged now. Take a look when you're free, thanks.

0xNeshi commented on lib/crypto/src/test_helpers.rs:
```diff
@@ -1,5 +1,9 @@
 
+/// Creates a proptest strategy for non-empty vectors of random bytes.
+///
+/// Useful for testing cryptographic primitives that require non-empty inputs.
```
```suggestion
```

markopiers commented:
hey @0xNeshi thanks for the review, I squashed my commits down to one, also fixed the formatting errors from [nightly/fmt](https://github.com/OpenZeppelin/rust-contracts-stylus/actions/runs/14571750435/job/40870212267?pr=626) ci test failure from the previous run. Please approve the test again for this commit, it should pass now. Also, @qalisander please have a look at the changes I've made when you're free. Thanks guys.

markopiers commented:
@0xNeshi @qalisander could anyone please suggest me how to get the [check-publish / Check publish on crates.io (pull_request)](https://github.com/OpenZeppelin/rust-contracts-stylus/actions/runs/14572123102/job/40918063221?pr=626) test to pass? Although I'm not sure if the test is failing due to any of the changes I made.

bidzyyys reviewed (commented):
Looks good for me, should be checked by @qalisander.

qalisander reviewed (approved):
LGTM

markopiers commented:
@qalisander can this be merged now?
